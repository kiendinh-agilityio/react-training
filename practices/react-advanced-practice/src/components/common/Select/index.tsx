import { ChangeEvent, forwardRef, memo } from 'react';

// Import radix
import { Flex, Box } from '@radix-ui/themes';

// Import components
import { ArrowSelectIcon } from '@/components/common/Icons';

type OptionsType = { text: string; value: string }[];

interface SelectProps {
  label?: string;
  name: string;
  optionsList: OptionsType;
  value?: string;
  errorMessage?: string;
  onBlur?: (e: ChangeEvent<HTMLSelectElement>) => void;
  onChange: (e: ChangeEvent<HTMLSelectElement>) => void;
}

const Select = memo(
  forwardRef<HTMLSelectElement, SelectProps>(
    (
      { name, optionsList, label, value, onChange, errorMessage, onBlur }: SelectProps,
      ref,
    ) => {
      const renderOptions = (options: OptionsType) =>
        options.map((option) => (
          <option key={option.value} value={option.value}>
            {option.text}
          </option>
        ));

      return (
        <Flex direction="column" className="gap-1.5 text-dark font-regular">
          {label && <label>{label}</label>}
          <Flex className="relative">
            <select
              className={`appearance-none px-5 py-[15px] border w-full rounded-[15px] focus:outline-none mb-1.5 ${errorMessage ? 'border-danger' : 'border-input'}`}
              name={name}
              id={name}
              value={value}
              onChange={onChange}
              onBlur={onBlur}
              ref={ref}
            >
              {renderOptions(optionsList)}
            </select>
            <Box as="span" className="absolute bottom-1/2 right-[20px]">
              <ArrowSelectIcon />
            </Box>
          </Flex>
        </Flex>
      );
    },
  ),
);

export default Select;
