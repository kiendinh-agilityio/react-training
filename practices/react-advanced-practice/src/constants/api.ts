// Create api from json server auth
export const API_AUTH_URL: string = process.env.VITE_API_AUTH_URL || '';

// Create api from mockapi
export const BASE_API: string = process.env.VITE_API_URL || '';
