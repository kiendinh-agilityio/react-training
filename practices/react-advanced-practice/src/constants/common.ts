// Variables Colors Logo
export const COLORS_LOGO = {
  PRIMARY: '#2d3748',
  SECONDARY: '#fff',
};

// Variables for option Roles
export const ROLES = [
  {
    text: 'Manager',
    value: 'Manager',
  },
  {
    text: 'Programmer',
    value: 'Programmer',
  },
  {
    text: 'Executive',
    value: 'Executive',
  },
  {
    text: 'Designer',
    value: 'Designer',
  },
];
