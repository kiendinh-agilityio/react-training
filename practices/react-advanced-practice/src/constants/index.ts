export * from './common';
export * from './sidebar-items';
export * from './navbar';
export * from './messages';
export * from './regex';
export * from './api';
