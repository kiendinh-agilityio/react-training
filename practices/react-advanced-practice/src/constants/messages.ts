export const MESSAGE_ERROR = {
  REQUIRED_ERROR: (field: string) => `${field} is required`,
  INVALID_EMAIL: 'Email is invalid',
  INVALID_PASSWORD:
    'Password must have at least 8 characters, 1 special character and 1 number. Example@123',
  INVALID_NAME: 'Name must start with a capital letter, have no special characters',
  INVALID_SIGNIN:
    "The email ord password that you've entered is incorrect. Please try again.",
  SIGNIN_FAILED: 'An error occurred while signing in. Please try again later.',
  SIGNUP_FAILED: 'Something went wrong. Please try again.',
};
