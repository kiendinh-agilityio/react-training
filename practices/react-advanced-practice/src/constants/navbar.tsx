// Import components icon
import {
  DashboardIcon,
  UserIcon,
  LoginIcon,
  RegisterIcon,
} from '@/components/common/Icons';

// Create data for header navbar
export const HEADER_NAVBAR = [
  { label: 'DASHBOARD', href: '/dashboard', icon: <DashboardIcon /> },
  { label: 'PROFILE', href: '/profile', icon: <UserIcon /> },
  { label: 'SIGN UP', href: '/signup', icon: <LoginIcon /> },
  { label: 'SIGN IN', href: '/signin', icon: <RegisterIcon /> },
];

// Create data for footer navbar
export const FOOTER_NAVBAR = [
  { label: 'Creative Tim', href: 'https://www.creative-tim.com' },
  { label: 'Simmmple', href: 'https://www.simmmple.com' },
  { label: 'Blog', href: '/blog' },
  { label: 'License', href: '/license' },
];
