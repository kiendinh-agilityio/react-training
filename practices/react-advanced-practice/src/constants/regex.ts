export const REGEX = {
  EMAIL: /^[^\s@]+@[^\s@]+\.[a-zA-Z]{2,6}$/,
  PASSWORD: /^(?=[A-Z])(?=.*\d)(?=.*[!@#$%^&*(),.?":{}|<>])(?=.*[a-zA-Z]).{8,}$/,
  NAME: /^[A-Z][a-z]*(\s[A-Z][a-z]*)*$/,
};
