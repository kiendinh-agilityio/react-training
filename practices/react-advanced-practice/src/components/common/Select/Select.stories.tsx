import Select from '.';
import { ROLES } from '@/constants';

export default {
  component: Select,
  title: 'Select',
  tags: ['autodocs'],
  parameters: {
    layout: 'centered',
  },
};

export const SelectRoles = {
  args: {
    label: 'Roles',
    name: 'roles',
    value: 'roles',
    optionsList: ROLES,
  },
};
