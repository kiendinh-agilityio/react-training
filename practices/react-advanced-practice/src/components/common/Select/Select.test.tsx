import { render } from '@testing-library/react';
import Select from '.';

// Import constants
import { ROLES } from '@/constants';

// Mock ArrowSelectIcon component
jest.mock('@/components/common/Icons', () => ({
  ArrowSelectIcon: () => <span>â–¼</span>,
}));

describe('Select Component - Render', () => {
  it('renders correctly with label and options', () => {
    const { asFragment } = render(
      <Select
        label="Select an option"
        name="select-option"
        optionsList={ROLES}
        value="1"
        onChange={() => {}}
      />,
    );

    // Snapshot test to ensure proper rendering
    expect(asFragment()).toMatchSnapshot();
  });

  it('renders without label', () => {
    const { asFragment } = render(
      <Select name="select-option" optionsList={ROLES} value="1" onChange={() => {}} />,
    );

    // Snapshot test to ensure proper rendering without label
    expect(asFragment()).toMatchSnapshot();
  });
});
