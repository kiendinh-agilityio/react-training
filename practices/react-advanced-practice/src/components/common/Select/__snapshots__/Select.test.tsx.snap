// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Select Component - Render renders correctly with label and options 1`] = `
<DocumentFragment>
  <div
    class="rt-Flex flex-col gap-1.5 text-dark font-regular"
  >
    <label>
      Select an option
    </label>
    <div
      class="rt-Flex relative"
    >
      <select
        class="appearance-none px-5 py-[15px] border w-full rounded-[15px] focus:outline-none mb-1.5 border-input"
        id="select-option"
        name="select-option"
      >
        <option
          value="Manager"
        >
          Manager
        </option>
        <option
          value="Programmer"
        >
          Programmer
        </option>
        <option
          value="Executive"
        >
          Executive
        </option>
        <option
          value="Designer"
        >
          Designer
        </option>
      </select>
      <span
        class="absolute bottom-1/2 right-[20px]"
      >
        <span>
          ▼
        </span>
      </span>
    </div>
  </div>
</DocumentFragment>
`;

exports[`Select Component - Render renders without label 1`] = `
<DocumentFragment>
  <div
    class="rt-Flex flex-col gap-1.5 text-dark font-regular"
  >
    <div
      class="rt-Flex relative"
    >
      <select
        class="appearance-none px-5 py-[15px] border w-full rounded-[15px] focus:outline-none mb-1.5 border-input"
        id="select-option"
        name="select-option"
      >
        <option
          value="Manager"
        >
          Manager
        </option>
        <option
          value="Programmer"
        >
          Programmer
        </option>
        <option
          value="Executive"
        >
          Executive
        </option>
        <option
          value="Designer"
        >
          Designer
        </option>
      </select>
      <span
        class="absolute bottom-1/2 right-[20px]"
      >
        <span>
          ▼
        </span>
      </span>
    </div>
  </div>
</DocumentFragment>
`;
